pkg/olog/olog.go

```diff
@@ -0,0 +1,107 @@
+// Copyright 2023 Outreach Corporation. All Rights Reserved.
+
+// Description: Implements the public API for the olog package.
+
+// Package olog implements a lightweight logging library built around
+// the slog package. It aims to never mask the core slog.Logger type by
+// default. Provided is a global system for controlling logging levels
+// based on the package and module that a logger was created in, with a
+// system to update the logging level at runtime.
+//
+// This package does not provide the ability to ship logs to a remote
+// server, instead a logging collector should be used.
+package olog
+
+import (
+	"fmt"
+	"log/slog"
+
+	"github.com/getoutreach/gobox/pkg/callerinfo"
+)
+
+// New creates a new slog instance that can be used for logging. The
+// provided logger use the global handler provided by this package. See
+// the documentation on the 'handler' global for more information.
+//
+// The logger will be automatically associated with the module and
+// package that it was instantiated in. This is done by looking at the
+// call stack.
+//
+// Note: As mentioned above, this logger is associated with the module
+// and package that created it. So, if you pass this logger to another
+// module or package, the association will NOT be changed. This
+// includes the caller metadata added to every log line as well as
+// log-level management. If a type has a common logging format that the
+// other module or package should use, then a slog.LogValuer should be
+// implemented on that type instead of passing a logger around. If
+// trying to set attributes the be logged by default, this is not
+// supported without retaining the original association.
+func New() *slog.Logger {
```

Comment:
Should we preemptively pass functional options to this? Maybe even a context too? Just so we don't have to potentially change function signatures in the future. I guess functional options wouldn't change the signature anyways since those are usually variadic..

But another thing I wanted to call out was the inability to force format. Right now you can't override text/JSON format, it only decides based on environment. We should probably allow this to be overridable.

Reply:
Overriding the default handler would be good! I'll add that.

Reply:
I think we should hold off on options because `New()` is meant to not be configured right now. Given we can add it without breaking anything later, I should we should hold off until we have a good reason to do so.

Reply:
As for `NewWithHandler`, that doesn't play into this much because that's very much so meant to be used only for tests (honestly, we might want to unexport it).